src/views.rs

```diff
@@ -0,0 +1,36 @@
+use crate::server;
+use actix_web::error;
+use actix_web::{web, HttpResponse, Responder};
+use serde_derive::Deserialize;
+use std::sync::{Arc, Mutex};
+
+#[derive(Deserialize)]
+pub struct FireplaceFormData {
+    status: String,
+}
+
+pub async fn state_handler(
+    form: web::Form<FireplaceFormData>,
+    data: web::Data<Arc<Mutex<server::Data>>>,
+) -> impl Responder {
+    let (fan, flame) = match form.status.as_str() {
+        "allOn" => (true, true),
+        "fireplaceOn" => (false, true),
+        "allOff" => (false, false),
+        _ => {
+            return Err(error::ErrorBadRequest(format!(
+                "unknown status: {}",
+                form.status
+            )))
+        }
+    };
+    let fireplace_state = data.lock().unwrap();
```

Comment:
an option to eliminate unwraps: 

```rust
data.lock()
    .map_err(|_| error::ErrorInternalServerError("Internal Server Error"))
    .and_then(|fireplace_state| {
        fireplace_state.fireplace.lock()
            .map_err(|_| error::ErrorInternalServerError("Internal Server Error"))
            .and_then(|mut fireplace_controller| {
                fireplace_controller.set(fan, flame)
                    .map_err(|e| error::ErrorBadRequest(format!("Unable to set fireplace state: {}", e)))
                    .map(|_| HttpResponse::Ok())
            })
    })

```

Reply:
Nice, I had a lot of fun with this. I actually ended up just wiring the whole function into a map/map_err with the suggested enum change.